src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs

```diff
@@ -11,108 +11,178 @@
-            _packagesLock = new object();
-            _packages = new Dictionary<PackageType, PushedPackage>();
+            _testSettings = TestSettings.CreateFromEnvironment();
+            _pushLock = new SemaphoreSlim(initialCount: 1);
+            _packages = new Dictionary<PackageType, Package>();
         }
         
-        public async Task<Package> PushAsync(PackageType packageType, ITestOutputHelper logger)
+        public async Task<Package> PushAsync(PackageType requestedPackageType, ITestOutputHelper logger)
         {
-            PushedPackage pushedPackage;
-            lock (_packagesLock)
+            var pushedPackage = GetPushedPackage(requestedPackageType, logger);
+            if (pushedPackage != null)
             {
-                if (!_packages.TryGetValue(packageType, out pushedPackage))
-                {
-                    var package = CreatePackage(packageType);
-                    pushedPackage = new PushedPackage(package);
-                    _packages[packageType] = pushedPackage;
-                }
+                return pushedPackage;
             }
 
-            if (pushedPackage.Pushed)
-            {
-                logger.WriteLine($"Package {pushedPackage} push has already been pushed.");
-                return pushedPackage.Package;
-            }
+            return await PushUnpushedPackagesAsync(requestedPackageType, logger);
+        }
 
-            var acquired = await pushedPackage.PushLock.WaitAsync(0);
+        private async Task<Package> PushUnpushedPackagesAsync(PackageType requestedPackageType, ITestOutputHelper logger)
+        {
+            var acquired = await _pushLock.WaitAsync(0);
             try
             {
                 if (!acquired)
                 {
-                    logger.WriteLine($"Another test is already pushing {pushedPackage}. Waiting.");
-                    await pushedPackage.PushLock.WaitAsync();
-                    acquired = true;
+                    logger.WriteLine("Another test is already pushing the packages. Waiting.");
+                    await _pushLock.WaitAsync();
+                }
+
+                // Has the package already been pushed?
+                var pushedPackage = GetPushedPackage(requestedPackageType, logger);
+                if (pushedPackage != null)
+                {
+                    return pushedPackage;
                 }
 
-                if (pushedPackage.Pushed)
+                var packageTypes = GetPackageTypes(requestedPackageType);
+
+                // Push all of the package types that have not been pushed yet.
+                var pushTasks = new Dictionary<PackageType, Task<Package>>();
+                foreach (var packageType in packageTypes)
                 {
-                    logger.WriteLine($"Package {pushedPackage} has already been pushed.");
-                    return pushedPackage.Package;
+                    if (!_packages.TryGetValue(packageType, out Package package))
```

Comment:
Not using `_packagesLock` here? I see `GetPushedPackage` uses it on read.

Can we just use `GetPushedPackage` and check if it returns null?